src/UndisclosedPassword.php

```
@@ -0,0 +1,177 @@
+    const HIBP_COUNT_BASE = 0;
+    const SHA1_LENGTH = 40;
+
+    const PASSWORD_BREACHED = 'passwordBreached';
+    const WRONG_INPUT = 'wrongInput';
+    const CONNECTION_FAILURE = 'connFail';
+    const UNKNOWN_ERROR = 'unknownError';
+
+    protected $messageTemplates = [
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
+        self::WRONG_INPUT => 'The provided password failed to be correctly hashed, please verify your input',
+        self::CONNECTION_FAILURE => 'Unable to reach HIBP service, please try again later',
+        self::UNKNOWN_ERROR => 'Something happened beyond our control, error reporting should give more details',
+    ];
+
+    /**
+     * @var ClientInterface
+     */
+    private $httpClient;
+
+    /**
+     * @var RequestFactoryInterface
+     */
+    private $httpRequest;
+
+    /**
+     * @var ResponseFactoryInterface
+     */
+    private $httpResponse;
+
+    /**
+     * @var int
+     */
+    private $count = 0;
+
+    /**
+     * PasswordBreach constructor.
+     *
+     * @param ClientInterface $httpClient
+     * @param RequestFactoryInterface $httpRequest
+     * @param ResponseFactoryInterface $httpResponse
+     */
+    public function __construct(
+        ClientInterface $httpClient,
+        RequestFactoryInterface $httpRequest,
+        ResponseFactoryInterface $httpResponse
+    ) {
+        $this->httpClient = $httpClient;
+        $this->httpRequest = $httpRequest;
+        $this->httpResponse = $httpResponse;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function isValid($value)
+    {
+        if (! is_string($value)) {
+            $this->error(self::WRONG_INPUT);
+            return false;
+        }
+        try {
+            $isPwnd = $this->isPwnedPassword($value);
+        } catch (Exception\RuntimeException $runtimeException) {
+            $this->error(self::CONNECTION_FAILURE);
+            return false;
+        } catch (\Exception $exception) {
```

I don't think we should catch `\Exception` here: something went really wrong here, maybe in service wiring, so we should let the exception bubble up instead, making developers aware of the problem